source/common/upstream/maglev_lb.h

```diff
@@ -62,9 +64,81 @@ class MaglevTable : public ThreadAwareLoadBalancerBase::HashingLoadBalancer,
+                                    double max_normalized_weight, bool use_hostname_for_hashing);
+
   const uint64_t table_size_;
-  std::vector<HostConstSharedPtr> table_;
   MaglevLoadBalancerStats& stats_;
+
+private:
+  /**
+   * Implementation specific construction of data structures to represent the
+   * Maglev Table.
+   */
+  virtual void constructImplementationInternals(std::vector<TableBuildEntry>& table_build_entries,
+                                                double max_normalized_weight) PURE;
+
+  /**
+   * Log each entry of the maglev table (useful for debugging).
+   */
+  virtual void logMaglevTable(bool use_hostname_for_hashing) const PURE;
+};
+
+/**
+ * This is an implementation of Maglev consistent hashing that directly holds
+ * pointers.
+ */
+class OriginalMaglevTable : public MaglevTable {
+public:
+  OriginalMaglevTable(const NormalizedHostWeightVector& normalized_host_weights,
+                      double max_normalized_weight, uint64_t table_size,
+                      bool use_hostname_for_hashing, MaglevLoadBalancerStats& stats)
+      : MaglevTable(table_size, stats) {
+    constructMaglevTableInternal(normalized_host_weights, max_normalized_weight,
+                                 use_hostname_for_hashing);
+  }
+  virtual ~OriginalMaglevTable() = default;
+
+  // ThreadAwareLoadBalancerBase::HashingLoadBalancer
+  HostConstSharedPtr chooseHost(uint64_t hash, uint32_t attempt) const override;
+
+private:
+  void constructImplementationInternals(std::vector<TableBuildEntry>& table_build_entries,
+                                        double max_normalized_weight) override;
+
+  void logMaglevTable(bool use_hostname_for_hashing) const override;
+  std::vector<HostConstSharedPtr> table_;
+};
+
+/**
+ * This maglev implementation leverages the number of hosts to more efficiently
+ * populate the maglev table.
+ * TODO(kbaichoo): Re-evaluate whether we should have the abstraction on the
+ * table representation.
+ */
+class CompactMaglevTable : public MaglevTable {
+public:
+  CompactMaglevTable(const NormalizedHostWeightVector& normalized_host_weights,
+                     double max_normalized_weight, uint64_t table_size,
+                     bool use_hostname_for_hashing, MaglevLoadBalancerStats& stats);
+  virtual ~CompactMaglevTable() = default;
+
+  // ThreadAwareLoadBalancerBase::HashingLoadBalancer
+  HostConstSharedPtr chooseHost(uint64_t hash, uint32_t attempt) const override;
+
+private:
+  void constructImplementationInternals(std::vector<TableBuildEntry>& table_build_entries,
+                                        double max_normalized_weight) override;
+  void logMaglevTable(bool use_hostname_for_hashing) const override;
+
+  // Leverage a BitArray to more compactly fit represent the MaglevTable.
+  // The BitArray will index into the host_table_ which will provide the given
+  // host to load balance to.
+  BitArray table_;
```

Comment:
I have another suggestion which might further improve performance and reduce memory. Maybe for a follow-up.

At a high level, this could be a variable-size data structure. See `InlineStorage` in source/common/common/utility.h. It might break the abstraction between BitArray and CompactMaglevTable so it would be a little messy.

But at the end you'd have one less indirection and 8 fewer bytes: the unique_ptr in BitArray would not be needed as the data would be inline.

Reply:
Cool to learn about `InlineStorage` as an available technique. I think here the odds of it helping might be slim as the size of the BitArray object itself would balloon in size and wouldn't fit in cache lines. Which would likely lead to an additional memory access. We can revisit if this becomes an issue.